CMakeLists.txt

```diff
@@ -0,0 +1,139 @@
+# Copyright 2023 Viam Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# TODO: [required] libviam_rust_utils handling
+# TODO: [required] Generate and install .cmake and .pc for consumers
+# TODO: [required] Dockerfile
+# TODO: Examples should build *against* the installed SDK, not as part of it.
+# TODO: Runtime paths and make BUILD_SHARED_LIBS work
+# TODO: Check required versions of third party libraries
+# TODO: Compiler minima
+# TODO: Finding external projects vs building in (e.g grpc)
+# TODO: Better clang-format integration (versions, dependencies, only changed, etc.)
+# TODO: Version from git describe?
+# TODO: Export macros and visibilty annotations
+
+
+# This is a tricky decision. If we set this too old, we lose access to
+# modern CMake facilities that we want to be able to use, like file
+# sets. If we set it too new, then the system CMake can't be used to
+# build the product. Here, we make an opinionated choice to require
+# the very latest stable CMake as of this writing, arguing that for
+# most uses of this product, it is expected to be cross-compiling for
+# the target platform on a development system, meaning that we are not
+# constrained by the version of CMake available on target systems.
+cmake_minimum_required(VERSION 3.25 FATAL_ERROR)
```

Comment:
So, just realized that Debian 11 (stable/buster) which is what the raspberry Pi OS (as well as our own Canon build environment) is based on, only provides cmake 3.18. Ubuntu 22.04 only has 3.22. I would definitely ask that this be reduced to 3.18 if possible. We don't want to exclude two major distros and our primary platform from being able to use this easily.

Reply:
@Otterverse - 

Yes, I expected there might be some pushback here. As the comment says, I made an opinionated decision, and so I may well need to back out of it.

However, it is very easy to download and install a newer CMake binary for Linux (3.25 prebuilt binaries for x86_64/aarch64 linux are at https://github.com/Kitware/CMake/releases/tag/v3.25.2), so I'd argue that requiring a newer version than is offered by the target system(s) of record isn't a serious usability obstacle, and it will only be an obstacle until Debian Bookworm (eventually) works its way everywhere.

You can also get prebuilt CMake v3.25 for Debian 11 from bullseye-backports: https://packages.debian.org/bullseye-backports/cmake, and you can get CMake v3.25 for Ubuntu 22.04 from KitWare's apt repo: https://apt.kitware.com/. So I'm hopeful that we can mostly solve it with documentation?

I'm hoping we can stay on the more modern version here because there are a few very useful CMake features like `FILE_SET` that are hard to replicate and I believe we will lose if we downgrade to 3.[18,22]. That all can be worked around of course if the above arguments about the relative ease of obtaining modern cmake on older platforms are unconvincing to you. 

What I definitely do agree with along these same lines is that our toolchain (g++) and runtime dependency minima (grpc, boost, etc.) must be tied to what is available on Debian 11, and the SDK must build and function correctly using those toolchains and libraries. I'm just not sure we need to apply that same restriction to CMake where it is easy to get a newer version and it doesn't form part of the runtime.

Definitely interested in your thoughts.

Reply:
I didn't realize it was definitely using newer features. I'm used to software not having that many breaking changes between "minor" releases, and when I set it to 3.18 I thought I ran it and it worked, but... I think I ran it in the wrong terminal in a docker with a different copy, because it's definitely not working now when I try it.

I'm not the final authority here, but I'm okay sticking with 3.25 given the actual use of these features (especially as it's in backports for our main debian platforms.) I'd ask that the readme just get a note pointing users in the right direction for those sources.

Reply:
If we expect a non minority of users to run cmake, I don't think we should make them download another dependency. If that's the case we should drop the version a bit. If it's the minority, I think it's fine

Reply:
sure

Reply:
Resolving this comment on CMake 3.25 for now. We can revisit it as we gain feedback from users. If it becomes an obstacle we can always downgrade.